<a href="https://colab.research.google.com/github/nayoung718/essa_python_1/blob/main/%ED%8C%8C%EC%9D%B4%EC%8D%AC_%EB%A8%B8%EC%8B%A0%EB%9F%AC%EB%8B%9D_%EC%99%84%EB%B2%BD%EA%B0%80%EC%9D%B4%EB%93%9C_ch2_1_~_4_pg_87~117_%ED%95%84%EC%82%AC.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##파이썬 머신러닝 완벽가이드 ch2. 1 ~ 4 pg. 87~117 필사

CHAPTER 02 사이킷런으로 시작하는 머신러닝

1. 사이킷런 소개와 특징

-파이썬 기반 다른 머신러닝 패키지도 사이킷런 스타일의 API를 지향할 정도로 쉽고 가장 파이썬스러운 API 제공

-머신러닝 위한 다양한 알고리즘과 개발 위한 편리한 프레임 워크와 API 제공

-검증됐고, 많은 환경에서 사용되는 성숙한 라이브러리

In [3]:
pip install scikit-learn

In [4]:
import sklearn
print(sklearn.__version__)

1.2.2


2. 첫 번째 머신러닝 만들어보기-붓꽃 품종 예측하기

-꽃잎 길이와 너비, 꽃받침 길이와 너비 피처 기반 꽃 품종 예측->품종 분류

-분류는 대표적 지도학습 방법으로 다양한 피처와 분류 결정값인 레이블 데이터로 모델 학습한 뒤, 테스트 데이터 세트에서 레이블 예측

In [5]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [7]:
import pandas as pd

#붓꽃 데이터 세트 로딩
iris=load_iris()

#iris.data는 Iris 데이터 세트에서 피처만으로 된 데이터를 numpy로 가지고 있음
iris_data=iris.data

#iris.target은 붓꽃 데이터 세트에서 레이블(결정 값) 데이터를 numpy로 가지ㅗㄱ 있음
iris_label=iris.target
print('iris target 값:',iris_label)
print('iris target 명:',iris.target_names)

#붓꽃 데이터 세트를 자세히 보기 위해 DataFrame으로 변환
iris_df=pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df.head(3)

iris target 값: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target 명: ['setosa' 'versicolor' 'virginica']


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   label  
0      0  
1      0  
2      0

-피처에는 sepal length, sepal width, petal length, petal width 존재하며 레이블은 0(setosa),1(versicolor),2(virginica)

-학습용 데이터와 테스트 데이터 분리

In [9]:
x_train, x_test, y_train, y_test = train_test_split(iris_data,iris_label, test_size=0.2, random_state=11)

-test_size는 전체 데이터 세트 중 테스트 데이터 세트 비율

-random_state는 호출할 때마다 같은 학습/테스트 용 데이터 세트 생성하기 위해 주어지는 난수 발생 값(random값 만드는 seed 느낌)

In [10]:
#DecisionTreeClassifier 객체 생성
dt_clf=DecisionTreeClassifier(random_state=11)

#학습 수행
dt_clf.fit(x_train,y_train)

DecisionTreeClassifier(random_state=11)

-DecisionTreeClassifier 객체 학습 데이터 기반 학습 완료, 이를 이용해 예측 수행: 학습 데이터 아닌 다른 데이터 이용해야함(테스트 데이터 세트)

In [11]:
#학습 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행.
pred=dt_clf.predict(x_test)

예측 결과 기반 DecisionTreeClassifier의 예측 성능 평가(정확도 평가)

In [12]:
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.933333


-예측 프로세스: 데이터 세트 분리(학습/테스트)-> 모델 학습-> 예측 수행-> 평가

3. 사이킷런의 기반 프레임워크 익히기

-Estimator 이해 및 fit(), predict() 메서드: fit() ML 모델 학습, predict() 학습된 모델 예측

-fit_transform(): 입력데이터 형태에 맞춰 데이터 변환하기 위한 사전 구조 맞추는 작업

-사이킷런의 주요 모듈, 내장된 예제 데이터 세트: 일반적 딕셔너리 형태, 키(data, target, target_name, feature_names, DESCR)

-붓꽃 데이터 세트 생성

In [13]:
from sklearn.datasets import load_iris

iris_data=load_iris()
print(type(iris_data))

<class 'sklearn.utils._bunch.Bunch'>


In [14]:
keys=iris_data.keys()
print('붓꽃 데이터 세트의 키들:',keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


In [15]:
print('\n feature_names의 type:',type(iris_data.feature_names))
print(' feature_names의 shape:',len(iris_data.feature_names))
print(iris_data.feature_names)

print('\n target_names의 type:',type(iris_data.target_names))
print(' target_names의 shape:',len(iris_data.target_names))
print(iris_data.target_names)

print('\n data의 type:',type(iris_data.data))
print(' data의 shape:',iris_data.data.shape)
print(iris_data['data'])

print('\n target의 type:',type(iris_data.target))
print(' target의 shape:',iris_data.target.shape)
print(iris_data.target)


 feature_names의 type: <class 'list'>
 feature_names의 shape: 4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names의 type: <class 'numpy.ndarray'>
 target_names의 shape: 3
['setosa' 'versicolor' 'virginica']

 data의 type: <class 'numpy.ndarray'>
 data의 shape: (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]
 [5

4. Model Selection 모듈 소개

-학습/테스트 데이터 세트 분리 - train_test_split()

In [16]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

iris=load_iris()
dt_clf=DecisionTreeClassifier()
train_data=iris.data
train_label=iris.target
dt_clf.fit(train_data,train_label)

#학습 데이터 세트로 예측 수행
pred=dt_clf.predict(train_data)
print('예측 정확도:',accuracy_score(train_label,pred))

예측 정확도: 1.0


-> 이미 학습한 학습 데이터 세트 기반 예측하여 예측 결과 100% : 학습용 데이터가 아닌 테스트 데이터 세트로 예측 수행해보아야함

-train_test_split() 이용해 테스트 데이터 세트를 전체 30%로, 학습 데이터 세트를 700%로 분리, random_state=121

In [17]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

dt_clf=DecisionTreeClassifier()
iris_data=load_iris()

x_train, x_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size=0.3, random_state=121)

-학습 데이터 기반 DecisionTreeClassifier 학습 후 이 모델 이용해 예측 정확도 측정

In [18]:
dt_clf.fit(x_train, y_train)
pred=dt_clf.predict(x_test)
print('예측 정확도: {0: .4f}'.format(accuracy_score(y_test, pred)))

예측 정확도:  0.9556


-> 붓꽃 데이터는 150개 데이터로 데이터 양이 크지 않아 여기서 30%인 테스트 데이터는 45개 밖에 안되어서 예측 성능 판단에는 그리 적절하지 않음

-교차 검증: 위의 방법에서 과적합 문제 가능하여 교차 검증 실행: 모의고사 여러번 보는 느낌으로 많은 학습과 검증 세트에서 알고리즘 학습과 평가 수행-> 별도의 여러 세트로 구성된 학습 데이터 세트와 검증 데이터 세트에서 학습과 평가 수행

-K 폴드 교차 검증: K개의 데이터 폴드 세트를 먼저 만들어서 K번 만큼 각 폴트 세트에 학습과 검증 평가 반복적 수행-> 학습 데이터 세트와 검증 데이터 세트 점진적 변경하면서 학습과 검증 수행

In [19]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris=load_iris()
features=iris.data
label=iris.target
dt_clf=DecisionTreeClassifier(random_state=156)

#5개의 폴드 세트로 분리하는 KFold 객체와 폴드 세트별 정확도를 담을 리스트 객체 생성.
kfold=KFold(n_splits=5)
cv_accuracy=[]
print('붓꽃 데이터 세트 크기:',features.shape[0])

붓꽃 데이터 세트 크기: 150


In [20]:
n_iter=0

In [21]:
#KFold 객체의 split()를 호출하면 폴드 별 학습용, 검증용 테스트의 로우 인덱스를 array로 반환
for train_index, test_index in kfold.split(features):
  #kfold.split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
  x_train, x_test=features[train_index],features[test_index]
  y_train,y_test=label[train_index],label[test_index]
  #학습 및 예측
  dt_clf.fit(x_train,y_train)
  pred=dt_clf.predict(x_test)
  n_iter+=1
  #반복 시마다 정확도 측정
  accuracy = np.round(accuracy_score(y_test,pred),4)
  train_size=x_train.shape[0]
  test_size=x_test.shape[0]
  print('\n#{0} 검증 세트 인덱스: {1}'.format(n_iter,test_index))
  cv_accuracy.append(accuracy)

#개별 iteration별 정확도를 합하여 평균 정확도 계산
print('\n## 평균 검증 정확도:',np.mean(cv_accuracy))


#1 검증 세트 인덱스: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

#2 검증 세트 인덱스: [30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

#3 검증 세트 인덱스: [60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

#4 검증 세트 인덱스: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

#5 검증 세트 인덱스: [120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도: 0.9


-5번 교차 검증 결과 평균 검증 정확도 0.9, 교차 검증시마다 검증 세트의 인덱스 달라짐 확인 가능

-Stratified K 폴드: 불균형한 분포도 가진 레이블 데이터 집합 위한 K 폴드 방식

In [22]:
import pandas as pd

iris=load_iris()
iris_df=pd.DataFrame(data=iris.data,columns=iris.feature_names)
iris_df['label']=iris.target
iris_df['label'].value_counts()

0    50
1    50
2    50
Name: label, dtype: int64

In [24]:
kfold=KFold(n_splits=3)
n_iter=0
for train_index, test_index, in kfold.split(iris_df):
  n_iter+=1
  label_train=iris_df['label'].iloc[train_index]
  label_test=iris_df['label'].iloc[test_index]
  print('## 교차 검증: {0}'.format(n_iter))
  print('학습 레이블 데이터 분포: \n',label_train.value_counts())
  print('검증 레이블 데이터 분포: \n',label_test.value_counts())
  iris_df['label'].value_counts()

## 교차 검증: 1
학습 레이블 데이터 분포: 
 1    50
2    50
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 0    50
Name: label, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포: 
 0    50
2    50
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 1    50
Name: label, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포: 
 0    50
1    50
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 2    50
Name: label, dtype: int64


-StratifiedKFold 사용 폴드세트 3개

In [25]:
from sklearn.model_selection import StratifiedKFold

skf=StratifiedKFold(n_splits=3)
n_iter=0

for train_index, test_index in skf.split(iris_df, iris_df['label']):
  n_iter+=1
  label_train=iris_df['label'].iloc[train_index]
  label_test=iris_df['label'].iloc[test_index]
  print('## 교차 검증: {0}'.format(n_iter))
  print('학습 레이블 데이터 분포: \n',label_train.value_counts())
  print('검증 레이블 데이터 분포: \n',label_test.value_counts())

## 교차 검증: 1
학습 레이블 데이터 분포: 
 2    34
0    33
1    33
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 0    17
1    17
2    16
Name: label, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포: 
 1    34
0    33
2    33
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 0    17
2    17
1    16
Name: label, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포: 
 0    34
1    33
2    33
Name: label, dtype: int64
검증 레이블 데이터 분포: 
 1    17
2    17
0    16
Name: label, dtype: int64


-StratifiedKFold 이용해 데이터 분리한 것으로 위에서 추출한 데이터 그대로 이용

In [27]:
dt_clf=DecisionTreeClassifier(random_state=156)

skfold=StratifiedKFold(n_splits=3)
n_iter=0
cv_accuracy=[]

#StratifiedKFold의 split() 호출 시 반드시 레이블 데이터 세트도 추가 입력 필요
for train_index, test_index in skfold.split(features, label):
  #split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
  x_train, x_test=features[train_index],features[test_index]
  y_train, y_test=label[train_index], label[test_index]
  #학습 및 예측
  dt_clf.fit(x_train, y_train)
  pred=dt_clf.predict(x_test)

  #반복 시마다 정확도 측정
  n_iter+=1
  accuracy=np.round(accuracy_score(y_test,pred),4)
  train_size=x_train.shape[0]
  test_size=x_test.shape[0]
  print('\n#{0} 교차 검증 정확도: {1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
  print('#{0} 검증 세트 인덱스: {1}'.format(n_iter, test_index))
  cv_accuracy.append(accuracy)

  #교차 검증별 정확도 및 평균 정확도 계산
  print('\n## 교차 검증별 정확도:',np.round(cv_accuracy,4))
  print('## 평균 검증 정확도:',np.mean(cv_accuracy))


#1 교차 검증 정확도: 0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#1 검증 세트 인덱스: [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]

## 교차 검증별 정확도: [0.98]
## 평균 검증 정확도: 0.98

#2 교차 검증 정확도: 0.94, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#2 검증 세트 인덱스: [ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 116 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132]

## 교차 검증별 정확도: [0.98 0.94]
## 평균 검증 정확도: 0.96

#3 교차 검증 정확도: 0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#3 검증 세트 인덱스: [ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  83  84
  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149]

## 교차 검증별 정확도: [0.98 0.94 0.98]
## 평균 검증 정확도: 0.9666666666666667


-3개의 stratified k 폴드로 교차 검증 결과 평균 검증 정확도 약 96.67% 로 측정, 왜곡된 레이블 데이터 세트 경우 반드시 stratified k 폴드 이용해 교차 검증해야 함

-교차 검증을 보다 간편하게 - cross_val_score(): KFold로 데이터 학습하고 예측하는 코드를 보며 폴드세터 설정, for루프에서 반복 학습 및 테스트 데이터의 인덱스 추출 후 반복적으로 학습과 예측 수행하고 예측 성능 반환 하는 과정 한꺼번에 수행해주는 API

In [28]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data=load_iris()
dt_clf=DecisionTreeClassifier(random_state=156)

data=iris_data.data
label=iris_data.target

#성능 지표는 정확도, 교차 검증 세트는 3개
scores=cross_val_score(dt_clf, data, label, scoring='accuracy',cv=3)
print('교차 검증별 정확도:',np.round(scores,4))
print('평균 검증 정확도:',np.round(np.mean(scores),4))

교차 검증별 정확도: [0.98 0.94 0.98]
평균 검증 정확도: 0.9667


-GridSearchCV - 교차 검증과 최적 하이퍼 파라미터 튜닝을 한번에: Classfier나 Regressor 같은 알고리즘에 사용되는 하이퍼 파라미터를 순차적으로 입력하면서 편리하게 최적의 파라미터 도출 방안 제공

In [29]:
grid_parameters={'max_depth':[1,2,3],'min_samples_split':[2,3]}

-GridSearchCV 클래스 생성자 주요 파라미터: estimator, param_grid, scoring, cv, refit

In [30]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import load_iris

#데이터 로딩하고 학습 데이터와 테스트 데이터 분리
iris=load_iris()
x_train, x_test, y_train, y_test= train_test_split(iris_data.data, iris_data.target, test_size=0.2, random_state=121)
dtree=DecisionTreeClassifier()

### 파라미터를 딕셔너리 형태로 설정
parameters={'max_depth':[1,2,3],'min_samples_split':[2,3]}

In [32]:
import pandas as pd

# param_grid의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정.
### refit=True가 default임. True면 가장 좋은 파라미터 설정으로 재학습시킴.
grid_dtree=GridSearchCV(dtree,param_grid=parameters, cv=3, refit=True)
#붓꽃 학습 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가
grid_dtree.fit(x_train, y_train)

#GridSearchCV 결과를 추출해 DataFrame으로 변환
scores_df=pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params','mean_test_score','rank_test_score','split0_test_score','split1_test_score','split2_test_score']]

params  mean_test_score  rank_test_score  \
0  {'max_depth': 1, 'min_samples_split': 2}         0.700000                5   
1  {'max_depth': 1, 'min_samples_split': 3}         0.700000                5   
2  {'max_depth': 2, 'min_samples_split': 2}         0.958333                3   
3  {'max_depth': 2, 'min_samples_split': 3}         0.958333                3   
4  {'max_depth': 3, 'min_samples_split': 2}         0.975000                1   
5  {'max_depth': 3, 'min_samples_split': 3}         0.975000                1   

   split0_test_score  split1_test_score  split2_test_score  
0              0.700                0.7               0.70  
1              0.700                0.7               0.70  
2              0.925                1.0               0.95  
3              0.925                1.0               0.95  
4              0.975                1.0               0.95  
5              0.975                1.0               0.95

-> 6개의 결과 볼 수 있고, 이는 하이퍼 파라미터 max_depth와 min_samples_split을 순차적으로 6번 변경하면서 학습 및 평가 수행했음을 나타냄

-params 칼럼에는 수행할 때마다 적용된 개별 하이퍼 파라미터값 나타냄

-rank_test_score는 하이퍼 파라미터별로 성능이 좋은 score 순위를 나타냄, 1이 가장 뛰어난 순위고 이때 파라미터가 최적의 하이퍼 파라미터

-mean_test_score는 개별 하이퍼 파라미터별로 CV의 폴딩 테스트 세트에 대해 총 수행한 평가 평균값

-> 최적 하이퍼 파리미터 값과 그때 정확도 알아보자

In [33]:
print('GridSearchCV 최적 파라미터:',grid_dtree.best_params_)
print('GridSearchCV 최고 정확도:{0:.4f}'.format(grid_dtree.best_score_))

GridSearchCV 최적 파라미터: {'max_depth': 3, 'min_samples_split': 2}
GridSearchCV 최고 정확도:0.9750


-이미 학습된 best_estimator_를 이용해 앞에서 train_test_split()으로 분리한 테스트 데이터 세트에 대해 예측하고 성능 평가

In [34]:
#GridSearchCV의 refit으로 이미 학습된 estimator 반환
estimator=grid_dtree.best_estimator_

#GridSearchCV의 best_estimator_는 이미 최적 학습이 됐으므로 별도 학습 필요 없음
pred=estimator.predict(x_test)
print('테스트 데이터 세트 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

테스트 데이터 세트 정확도: 0.9667
